### DISCUSSION TITLE
Postgresql or Mysql client export based on `DATABASE_URL`, typing issue.

### DISCUSSION THREAD

**Discussion by baptisteArno** [prisma/prisma#17713] [Q&A] [answered]
My project needs to be able to connect to either a PostgreSQL database or a MySQL database.

So I have 2 identical schemas that just output 2 different clients.

Now I just export the right client depending on the `DATABASE_URL` env variable:

```ts
import * as Postgresql from './prisma-clients/postgresql'
import * as Mysql from './prisma-clients/mysql'

export * from './prisma-clients/postgresql'

export const PrismaClient = process.env.DATABASE_URL?.includes('mysql')
  ? Mysql.PrismaClient
  : Postgresql.PrismaClient
```

But then the typing isn't right when I try to execute any operation. For example:

```ts
const prismaGlobal = global as typeof global & {
  prisma?: PrismaClient
           //^^^^^^^ 'PrismaClient' refers to a value, but is being used as a type here. Did you mean 'typeof PrismaClient'?
}
```

Or

```ts
await prisma.publicTypebot.findMany({
                          //^^^^^^ This expression is not callable. Each member of the union type '(<T extends PublicTypebotFindManyArgs>(args?: SelectSubset<T, PublicTypebotFindManyArgs> | undefined) => PrismaPromise<...>) | (<T extends PublicTypebotFindManyArgs>(args?: SelectSubset<...> | undefined) => PrismaPromise<...>)' has signatures, but none of those signatures are compatible with each other.
    // ...
  })
```

Both clients should have similar typings, right?

**Comment by baptisteArno** [comment]
A solution could be to cast the client like this for example:

```ts
(process.env.DATABASE_URL?.includes('mysql')
  ? Mysql.PrismaClient
  : Postgresql.PrismaClient) as unknown as typeof Mysql.PrismaClient
```

But not clean.

**Comment by millsp** [comment, marked as answer]
Hey Baptiste. I think what you are trying is possible via casting as an intersection. That would be something like `as unknown as typeof Mysql.PrismaClient & Postgresql.PrismaClient`. While that will work, you may run into a slowed down IDE experience. Since `mysql` is a clear subset of `postgresql`, I would suggest to avoid that and just reuse the types from `mysql`.

So why is PostgreSQL and MySQL types incompatible? Basically, PostgreSQL has a slightly wider API because we expose a mode for [case sensitivity](https://www.prisma.io/docs/concepts/components/prisma-client/filtering-and-sorting) on filtering. That might be a surprise to you at runtime, as Postgres and Mysql have different defaults, so I think it's worth mentioning.

So I suggest that you choose the smaller subset, which is to cast as MySQL like you've shown above.

**Reply by baptisteArno** [reply]
Alright, thank you so much for the info 🙏

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
